### PR TITLE
feat: keep_names for FunctionDeclaration

### PR DESCRIPTION
**Pull request by IWANABETHATGUY**

### CODE REVIEW

File: crates/rolldown/tests/esbuild/default/dynamic_import_with_expression_cjs/artifacts.snap [hunk 1]
```diff
@@ -1,5 +1,6 @@
 ---
 source: crates/rolldown_testing/src/integration_test.rs
+snapshot_kind: text
```
Comment (hyf0) [inline crates/rolldown/tests/esbuild/default/dynamic_import_with_expression_cjs/artifacts.snap]: Try delete all changed snapshots and do a update again. This should remove unchanged snapshots.

File: crates/rolldown/src/module_finalizers/scope_hoisting/impl_visit_mut.rs [hunk 2]
```diff
@@ -844,4 +879,24 @@ impl<'ast> ScopeHoistingFinalizer<'_, 'ast> {
...
+
+  fn process_fn_decl(&mut self, decl: &mut allocator::Box<'_, ast::Function<'_>>) -> Option<()> {
+    if !self.ctx.options.keep_names {
+      return None;
+    }
+    let id = decl.id.as_ref()?;
+    let symbol_id = id.symbol_id.get()?;
+    let symbol_ref = (self.ctx.id, symbol_id).into();
+    dbg!(&self.ctx.renamed_symbol_map);
+    if let Some(original_name) = self.ctx.renamed_symbol_map.get(&symbol_ref) {
```
Comment (hyf0) [inline crates/rolldown/src/module_finalizers/scope_hoisting/impl_visit_mut.rs]: Does this do the same work as `renamed_symbol_map`? 
```rust
let is_changed = symbol_ref.name() == self.ctx.symbol_db.canonical_name_for(symbol_ref);
```

File: crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs [hunk 3]
```diff
@@ -264,6 +263,7 @@ impl GenerateStage<'_> {
             depended_symbols.insert(self.link_output.runtime.resolve_symbol("__toCommonJS"));
             depended_symbols.insert(entry.namespace_object_ref);
           }
+          // depended_symbols.insert(self.link_output.runtime.resolve_symbol("__name"));
```
Comment (hyf0) [inline crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs]: I guess this should be done in `reference_needed_symbols`. Insert reference to `__name` for `StmtInfo` that contains fn or class.

Comment (IWANABETHATGUY) [reply crates/rolldown/src/stages/generate_stage/compute_cross_chunk_links.rs]: then you need to insert every function and class for  `__name`

File: crates/rolldown/src/module_finalizers/scope_hoisting/impl_visit_mut.rs (hunk 2 above)
Comment (IWANABETHATGUY) [reply crates/rolldown/src/module_finalizers/scope_hoisting/impl_visit_mut.rs]: the same effect but should be faster, `canonical_name_for` needs two steps, first get the finalized symbolRef, then get the deconflicted name, get symbol also need extra lookup 1 step vs 3 step.

File: crates/rolldown/tests/esbuild/default/dynamic_import_with_expression_cjs/artifacts.snap (hunk 1 above)
Comment (IWANABETHATGUY) [reply crates/rolldown/tests/esbuild/default/dynamic_import_with_expression_cjs/artifacts.snap]: This pr is not finished yet, I will clean up in final step.

**Review by hyf0 — APPROVED** [review]
No serious blocking points. LGTM.

**Comment by IWANABETHATGUY** [comment]
* **#3023** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/3023?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3021** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/3021?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/3021?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by IWANABETHATGUY** [comment]
The benchmark looks good now.

**Pull request merged**